proto2ros/proto2ros/output/templates/conversions.cpp.jinja

```diff
@@ -0,0 +1,396 @@
+// THIS IS AN AUTOGENERATED FILE. DO NOT EDIT THIS FILE DIRECTLY.
+// See conversions.cpp.jinja template in the proto2ros.output.templates Python module.
+
+#include "{{ package_name }}/conversions.hpp"
+
+#include <stdexcept>
+#include <sstream>
+#include <utility>
+
+#include <google/protobuf/any.h>
+#include <google/protobuf/map.h>
+
+#include <rclcpp/serialization.hpp>
+#include <rclcpp/serialized_message.hpp>
+
+#include <proto2ros/conversions.hpp>
+
+{#-
+  Expands to code for ROS to Protobuf field conversion.
+
+  Args:
+    source: identifier of the ROS message field to convert from.
+    destination: identifier of the Protobuf message field to convert to.
+    spec: annotated ROS message field specification.
+-#}
+{%- macro ros_to_proto_field_code(source, destination, spec) -%}
+    {%- if not spec.type.is_primitive_type() -%}
+        {%- set type_spec = known_message_specifications.get(to_ros_base_type(spec.type)) -%}
+        {%- if type_spec and type_spec.annotations.get("proto-class") == "enum" -%}
+            {#- Handle enum case i.e. extract integral values from ROS wrapper messages. -#}
+            {%- if spec.type.is_array -%}
+{{ destination }}->Reserve({{ source }}.size());
+for (const auto& item : {{ source }}) {
+    {{ destination }}->Add(static_cast<{{ type_spec.annotations["proto-type"] | as_pb2_cpp_type }}>(item.value));
+}
+            {%- else -%}
+{{ destination | rreplace("mutable", "set", 1) | rreplace("()", "", 1) }}(
```

Comment:
I'm not super proud of this but the C++ API for Protobuf message is so inconsistent across types that it hurts.